File: day68/LRUCache.cpp
// 146. LRU Cache
#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;

class LRUCache{
  public:
  class ListNode{
    public:
    int key, val;
    ListNode* next;
    ListNode* prev;
    ListNode(int k, int v): key(k), val(v), next(nullptr), prev(nullptr){};
  };
  ListNode* head;
  ListNode* tail;
  unordered_map<int, ListNode*> m;
  int capacity;

  LRUCache(int cap){
    capacity = cap;
    head = new ListNode(-1, -1);
    tail = new ListNode(-1, -1);
    head->next = tail;
    tail->prev = head;
  }

  void addNode(ListNode* newNode){
    ListNode* nextNode = head->next;
    
    newNode->next = nextNode;
    newNode->prev = head;

    head->next =newNode;
    nextNode->prev = newNode;
  };
  void removeNode(ListNode* targetNode){
    ListNode* prevNode = targetNode->prev;
    ListNode* nextNode = targetNode->next;
    
    prevNode->next = nextNode;
    nextNode->prev = prevNode;
  };

  void put(int key, int val){
    if(m.find(key) != m.end()){
      ListNode* existingNode = m[key];
      removeNode(existingNode);
      delete existingNode;
    }else{
      if(m.size() == capacity){
        ListNode* leastNode = tail->prev;
        m.erase(leastNode->key);
        removeNode(leastNode);
        delete leastNode;
      }
    }
    ListNode* newNode = new ListNode(key, val);
    addNode(newNode);
    m[key] = newNode;
  } 
  int get(int key){
    if(m.find(key) == m.end()) return -1;
    ListNode* accessedNode = m[key];
    removeNode(accessedNode);
    addNode(accessedNode);
    return accessedNode->val;
  }

  void displayElements(){
    for(auto it = m.begin(); it != m.end(); ++it){
      cout<<it->first<<" : "<<it->second->val<<" -> ";
    }
    cout<<"NULL"<<endl;
  }
};

int main(){
  LRUCache obj =  LRUCache(2);
  obj.put(1, 5);
  obj.put(7, 15);
  cout<<obj.get(1)<<endl;
  cout<<obj.get(7)<<endl;
  obj.displayElements();
  obj.put(8, 25);
  obj.displayElements();
  cout<<obj.get(7)<<endl;

  return 0;
}

File: day99/LRUCache.java
// 146. LRU Cache
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.Map.Entry;

public class LRUCache {
  // LinkedHashMap<Integer, Integer> m;
  // int cap;

  // public LRUCache(int capacity) {
  //   cap = capacity;
  //   m = new LinkedHashMap<>(capacity,  0.75f, true){
  //     @Override
  //     protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
  //       return size() > cap; // Auto remove eldest when capacity exceeded
  //     }
  //   };
  // }

  // public int get(int key) {
  //   return m.getOrDefault(key, -1);
  // }
  
  // public void put(int key, int value) {
  //   m.put(key, value);
  // }

  // public void printEntries(){
  //   for(Entry<Integer, Integer> e: m.entrySet()){
  //     System.out.println(e.getKey()+" "+e.getValue());
  //   }
  // }

  private static class ListNode{
    int key, val;
    ListNode next, prev;
    public ListNode(int k, int v){
      key = k;
      val = v;
    }
  }
  private int cap;
  private ListNode head, tail;
  private Map<Integer, ListNode> m;
  public LRUCache(int capacity){
    cap = capacity;
    head = new ListNode(0, 0);
    tail = new ListNode(0, 0);
    head.next = tail;
    tail.prev = head;
    m = new HashMap<>();
  }
  
  private void remove(ListNode node){
    node.prev.next = node.next;
    node.next.prev = node.prev;
  }
  private void insertAtHead(ListNode node){
    node.prev = head;
    node.next = head.next;
    head.next.prev = node;
    head.next = node;
  }
  public void put(int key, int value) {
    if(m.containsKey(key)){
      ListNode existingNode = m.get(key);
      existingNode.val = value;
      remove(existingNode);
      insertAtHead(existingNode);
    }else{
      ListNode newNode = new ListNode(key, value);
      insertAtHead(newNode);
      m.put(key, newNode);
      if(m.size() > cap){
        ListNode lru = tail.prev;
        m.remove(lru.key);
        remove(lru);
      }
    }
  }
  public int get (int key){
    if(m.containsKey(key)){
      ListNode existingNode = m.get(key);
      remove(existingNode);
      insertAtHead(existingNode);
      return existingNode.val;
    }
    return -1;
  }

  public static void main(String[] args) {
    LRUCache lc = new LRUCache(3);
    lc.put(1, 11);
    lc.put(2, 12);
    lc.put(3, 13);
    // lc.printEntries();
    lc.put(4, 14);
    System.out.println(lc.get(3));
    // lc.printEntries();
  }
}